### PR TITLE
deprecate open_on_setup

### LINKED ISSUE #1669: New Startup Configuration Mechanism Incoming
The existing startup mechanisms such an `open_on_setup` are problematic, with recent changes resulting in many regressions #1668.

The options are nonatomic and difficult (sometimes impossible) to configure for each use case.

There are often race conditions around startup and events with vim itself and other plugins.

A new mechanisms to specify startup behaviour with more granularity is under way.

The existing mechanism will be retained if an original configuration options such as `open_on_setup` is specified, however the user will be notified of the new mechanism on startup and pointed to a wiki page to explain migration to the new.

**Update 20230109:** the new mechanism will be a set of recipes for the user to control behaviour programatically, along with the events to allow this.

**Update 20230129:** existing configuration will be deprecated in favour of a self serve mechanism: [Open At Startup](https://github.com/nvim-tree/nvim-tree.lua/wiki/Open-At-Startup)

### PR DESCRIPTION
**Pull request by alex-courtis**
fixes #1669

Wiki with all the details: https://github.com/nvim-tree/nvim-tree.lua/wiki/Open-At-Startup

### CODE REVIEW

File: lua/nvim-tree.lua [hunk 1]
```diff
@@ -900,7 +901,17 @@ function M.setup(conf)
...
 
   vim.schedule(function()
-    M.on_enter(netrw_disabled)
+    if
+      #opts.ignore_ft_on_setup > 0
+      or opts.open_on_setup == true
+      or opts.open_on_setup_file
+      or opts.ignore_buffer_on_setup
+    then
+      notify.info "open_on_setup behaviour has been deprecated, please see https://github.com/nvim-tree/nvim-tree.lua/wiki/Open-At-Startup"
```
Comment (alex-courtis) [inline lua/nvim-tree.lua:910]: I suspect people might get upset, however I can't see any other way to do this.

**Review by gegoune — APPROVED** [review]

**Comment by alex-courtis** [comment]
"Open in window from previous session" longer occurs on startup. I'll prepare a PR in case anyone misses it.

**Comment by polyzen** [comment]
Shouldn't netrw be allowed to work now? Now attempting to open directories is treated as opening a file, as described in https://github.com/nvim-tree/nvim-tree.lua/issues/1956#issue-1563563124.

**Comment by alex-courtis** [comment]
> Shouldn't netrw be allowed to work now? Now attempting to open directories is treated as opening a file, as described in https://github.com/nvim-tree/nvim-tree.lua/issues/1956#issue-1563563124.

That's a really good point! netrw does have useful functionality including/excluding its "tree" handling.

With full control of the startup process, `:h nvim-tree.hijack_netrw` and `hijack_directories.enable` a user specific configuration that will work with netrw may be possible. We could come up with some [Recipes](https://github.com/nvim-tree/nvim-tree.lua/wiki/Recipes) and enhance the documentation around what is and is not possible.

I'd be grateful if you could explore this; some extra functionality may be required however we can address that as it arises.

**Pull request merged**